examples/bridge-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,452 @@
+static Device * gDevices[DYNAMIC_ENDPOINT_COUNT]; // number of dynamic endpoints count
+
+// Descriptor attribute storage on dynamic endpoint
+static uint8_t gDescriptorAttrStorage[DYNAMIC_ENDPOINT_COUNT][kDescriptorAttributeCount][kDescriptorAttributeArraySize];
+
+// (taken from chip-devices.xml)
+#define DEVICE_TYPE_CHIP_BRIDGE 0x0a0b
+// (taken from lo-devices.xml)
+#define DEVICE_TYPE_LO_ON_OFF_LIGHT 0x0100
+// Device Version for dynamic endpoints:
+#define DEVICE_VERSION_DEFAULT 1
+
+/* BRIDGED DEVICE ENDPOINT: contains the following clusters:
+   - On/Off
+   - Descriptor
+   - Bridged Device Basic
+   - Fixed Label
+*/
+
+// Declare On/Off cluster attributes
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_BEGIN(onOffAttrs)
+DECLARE_DYNAMIC_ATTRIBUTE(ZCL_ON_OFF_ATTRIBUTE_ID, BOOLEAN, 1, 0) /* on/off */
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_END(0x0001);
+
+// Declare Descriptor cluster attributes
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_BEGIN(descriptorAttrs)
+DECLARE_DYNAMIC_ATTRIBUTE(ZCL_DEVICE_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0),     /* device list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_SERVER_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0), /* server list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_CLIENT_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0), /* client list */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_PARTS_LIST_ATTRIBUTE_ID, ARRAY, kDescriptorAttributeArraySize, 0)   /* parts list */
+    DECLARE_DYNAMIC_ATTRIBUTE_LIST_END(0x0001);
+
+// Declare Bridged Device Basic information cluster attributes
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_BEGIN(bridgedDeviceBasicAttrs)
+DECLARE_DYNAMIC_ATTRIBUTE(ZCL_USER_LABEL_ATTRIBUTE_ID, CHAR_STRING, kUserLabelSize, 0), /* UserLabel */
+    DECLARE_DYNAMIC_ATTRIBUTE(ZCL_REACHABLE_ATTRIBUTE_ID, BOOLEAN, 1, 0)                /* Reachable */
+    DECLARE_DYNAMIC_ATTRIBUTE_LIST_END(0x0003);
+
+// Declare Fixed Label cluster attributes
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_BEGIN(fixedLabelAttrs)
+DECLARE_DYNAMIC_ATTRIBUTE(ZCL_LABEL_LIST_ATTRIBUTE_ID, ARRAY, kFixedLabelAttributeArraySize, 0) /* label list */
+DECLARE_DYNAMIC_ATTRIBUTE_LIST_END(0x0001);
+
+// Declare Cluster List for Bridged Light endpoint
+DECLARE_DYNAMIC_CLUSTER_LIST_BEGIN(bridgedLightClusters)
+DECLARE_DYNAMIC_CLUSTER(ZCL_ON_OFF_CLUSTER_ID, onOffAttrs), DECLARE_DYNAMIC_CLUSTER(ZCL_DESCRIPTOR_CLUSTER_ID, descriptorAttrs),
+    DECLARE_DYNAMIC_CLUSTER(ZCL_BRIDGED_DEVICE_BASIC_CLUSTER_ID, bridgedDeviceBasicAttrs),
+    DECLARE_DYNAMIC_CLUSTER(ZCL_FIXED_LABEL_CLUSTER_ID, fixedLabelAttrs) DECLARE_DYNAMIC_CLUSTER_LIST_END;
+
+// Declare Bridged Light endpoint
+DECLARE_DYNAMIC_ENDPOINT(bridgedLightEndpoint, bridgedLightClusters);
+
+int AddDeviceEndpoint(Device * dev, EmberAfEndpointType * ep, uint16_t deviceType)
+{
+    uint8_t index = 0;
+    while (index < DYNAMIC_ENDPOINT_COUNT)
+    {
+        if (NULL == gDevices[index])
+        {
+            gDevices[index] = dev;
+            EmberAfStatus ret;
+            while (1)
```

Comment:
what guarantees that this while loop returns? I.e. why do we know that emberAfSetDynamicEndpoint will eventually eitehr succeed or return 'duplicate exists'?

Some comments would be nice or alternatively a fixed size loop would also help.

Reply:
Removed infinite loop